server/src/main/java/org/apache/druid/server/coordinator/duty/CompactSegments.java

```diff
@@ -426,6 +426,17 @@ private CoordinatorStats doRun(
         if (config.getIoConfig() != null) {
           dropExisting = config.getIoConfig().isDropExisting();
         }
+        // force dropExisting to true if dropExisting is not set,
+        // dropExistingis false,
+        // or all segments to compact are tombstones.
+        // Otherwise, auto-compaction will cycle indefinitely since segments won't be compacted and their
+        // compaction state won't be set
```

Comment:
This comment is basically documentation but I fear that it is missing enough words to fully explain why we are forcing this setting even if the cluster wasn't configured to do this.  To someone without context, I fear they will get lost in the "if drop Existing is not set, dropExisting is false, all segments are tombstones" wondering what that means.   Maybe re-write as prose explaining the situation that this is attempting to resolve.  Something like

> if drop Existing is not set, dropExisting is false, all segments are tombstones, then it is possible for auto-compaction to cycle indefinitely because XYZ.  Thus, we force dropExisting to be true to work around this.

Then also throw in an extra note that this is working around a limitation in compaction and that we should look at how to make compaction be able to make progress even in cases like the one described here and when/if that is done, this workaround could be revisited.

Reply:
I will fix this in next commit. I have been thinking a little, not a lot, and it seems to me, at first sight, that auto-compaction should not track whether a segment was compacted in its payload. Somehow auto-compaction ought to reason about current state of segments and decide to compact them based on that state not whether they were compacted recently or not (i.e. this segment is tiny and there are other sibling segments in this time chunk that could be compacted together). But since state about compaction is saved in the segment payload and this state is needed to make the decision to compact we have to be able to save this state for all segments considered for auto-compaction and thus we need to force `dropExisting` to be able to compact those tombstones in this case. The semantics though should bot change. Previously there were only tombstones in this interval and after this there will be only tombstones as well (albeit with different segment granularity).

Reply:
Also, it looks like the code could be written without forcing dropExisting to true by entering the tombstone creation code in the sequential & parallel task by adding another condition in addition to `dropExisting`:  `ingestionSchema.getIOConfig().isDropExisting() || (storeCompactionState && pushed.getSegments().isEmpty()`. But the code will need to be updated in two places and it is also messy. Looks less messy if we just force the flag...

Reply:
Added comments as described (not we can also avoid forcing the flag as described above...please comment if you favor that other way instead)

Reply:
`storeCompactionState` should not be used. It can be set by the user and apply to non compaction task (i.e. batch ingestion reading data from s3).

Reply:
Ok, so the comments that I added should be the "fix" then thanks.